cmd/block-generator/runner/run.go

```diff
@@ -0,0 +1,260 @@
+		return filepath.Walk(args.Path, func(path string, info os.FileInfo, err error) error {
+			// Ignore the directory
+			if info.IsDir() {
+				return nil
+			}
+			runnerArgs := args
+			runnerArgs.Path = path
+			fmt.Printf("Running test for configuration '%s'\n", path)
+			return runnerArgs.run()
+		})
+	}
+
+	// Single file mode
+	return args.run()
+}
+
+func (r *Args) run() error {
+	// Start services
+	algodNet := fmt.Sprintf("localhost:%d", 11112)
+	indexerNet := fmt.Sprintf("localhost:%d", r.IndexerPort)
+	generatorShutdownFunc := startGenerator(r.Path, algodNet)
+	indexerShutdownFunc, err := startIndexer(r.IndexerBinary, algodNet, indexerNet, r.PostgresConnectionString)
+	if err != nil {
+		return fmt.Errorf("failed to start indexer: %w", err)
+	}
+
+	// Run the test, collecting results.
+	r.runTest(indexerNet, algodNet)
+
+	// Shutdown generator.
+	if err := generatorShutdownFunc(); err != nil {
+		return err
+	}
+
+	// Shutdown indexer
+	if err := indexerShutdownFunc(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Helper to record the import rate.
+func recordDataToFile(entry Entry, key string, out *os.File) error {
+	sum := 0.0
+	count := 0.0
+
+	for _, metric := range entry.Data {
+		var err error
+		if strings.HasPrefix(metric, "indexer_daemon_import_time_sec_sum") {
+			val := strings.Split(metric, " ")[1]
+			sum, err = strconv.ParseFloat(val, 64)
+		}
+		if strings.HasPrefix(metric, "indexer_daemon_import_time_sec_count") {
+			val := strings.Split(metric, " ")[1]
+			count, err = strconv.ParseFloat(val, 64)
+		}
+		if err != nil {
+			return fmt.Errorf("unable to parse metric '%s': %w", metric, err)
+		}
+	}
+	rate := sum / count
+
+	msg := fmt.Sprintf("%s:%f\n", key, rate)
+	if _, err := out.WriteString(msg); err != nil {
+		return fmt.Errorf("unable to write metric '%s': %w", key, err)
+	}
+	return nil
+}
+
+// Run the test for 'RunDuration', collect metrics and write them to the 'ReportDirectory'
+func (r *Args) runTest(indexerURL string, generatorURL string) error {
+	collector := &MetricsCollector{MetricsURL: fmt.Sprintf("http://%s/metrics", indexerURL)}
+
+	baseName := filepath.Base(r.Path)
+	baseNameNoExt := strings.TrimSuffix(baseName, filepath.Ext(baseName))
+	reportPath := path.Join(r.ReportDirectory, fmt.Sprintf("%s.report", baseNameNoExt))
+
+	report, err := os.Create(reportPath)
+	if err != nil {
+		return fmt.Errorf("unable to create report: %w", err)
+	}
+	defer report.Close()
+
+	// Run for r.RunDuration
+	start := time.Now()
+	for time.Since(start) < r.RunDuration {
+		time.Sleep(r.RunDuration / 10)
+		if err := collector.Collect("indexer_daemon_import_time_sec"); err != nil {
+			return fmt.Errorf("problem collecting metrics: %w", err)
+		}
+	}
+	if err := collector.Collect("indexer_daemon_import_time_sec"); err != nil {
+		return fmt.Errorf("problem collecting metrics: %w", err)
+	}
+
+	// Collect results.
+	durationStr := fmt.Sprintf("test_duration_seconds:%d\n", uint64(r.RunDuration.Seconds()))
```

Comment:
we should probably report the actual elapsed time

Reply:
done